### PR TITLE
x64: Add most remaining AVX lowerings

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit goes through `inst.isle` and adds a corresponding AVX lowering for most SSE lowerings. I opted to skip instructions where the SSE lowering didn't read/modify a register, such as `roundps`. I think that AVX will benefit these instructions when there's load-merging since AVX doesn't require alignment, but I've deferred that work to a future PR.

Otherwise though in this PR I think all (or almost all) of the 3-operand forms of AVX instructions are supported with their SSE counterparts. This should ideally improve codegen slightly by removing register pressure and the need for `movdqa` between registers. I've attempted to ensure that there's at least one codegen test for all the new instructions.

As a side note, the recent capstone integration into `precise-output` tests helped me catch a number of encoding bugs much earlier than otherwise, so I've found that incredibly useful in tests!

### CODE REVIEW

File: cranelift/codegen/src/isa/x64/inst/emit.rs
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/emit.rs:2005]: I'm not wild about the argument-dependent regalloc metadata shape here, and ideally I'd prefer us to split out variants like `XmmRmiRVexConst` that are used instead whenever the intent of an instruction is to produce a constant 0/1/whatever, and that assert the correct setup for that; but that's a separate refactor you don't necessarily need to do in this PR. Just noting as I see the existing pattern expanding...

**Review by cfallin — COMMENTED** [review]
Thank you for the really tedious and careful work here -- this will undoubtedly be a nice improvement for FP and SIMD code!

Two thoughts below: the first I think I do want before this PR goes in, the second is a note about a pattern needing a longer-term refactor that this PR extends (but it's pre-existing so I won't block on it here necessarily, unless you want to address).

File: cranelift/codegen/src/isa/x64/inst.isle
Comment (alexcrichton) [reply cranelift/codegen/src/isa/x64/inst.isle:241]: Sure, I was following [this precedent](https://github.com/bytecodealliance/wasmtime/blob/4fc768df36e53d1c9f42db6b3ea52af45a3dc129/cranelift/codegen/src/isa/x64/inst.isle#L356-L358) but I can split out too.

File: cranelift/codegen/src/isa/x64/inst/emit.rs
Comment (alexcrichton) [reply cranelift/codegen/src/isa/x64/inst/emit.rs:2005]: Sure yeah, I was just following the preexisting pattern for the xmm instructions, but I can try to update those too in a subsequent PR.

**Comment by alexcrichton** [comment]
Ok I think the two new commits should address the `produces_const` expansion in addition to the `Gpr`-vs-`Xmm` types. There's still more instructions that do conditional register allocation but I can try to tackle those separately.

**Review by cfallin — APPROVED** [review]
Fantastic, thanks for tackling both cleanups; getting rid of `produces_const` is really nice. LGTM!

**Comment by alexcrichton** [comment]
Local fuzzing has found an issue so taking this out of the merge queue.

**Comment by alexcrichton** [comment]
Turned out to be a benign mistake where I forgot to call `add_trap` for VEX-encoded instructions with memory operands. Before merging this though I'm going to let the fuzzer run overnight.

I've intentionally introduced an encoding bug where `vsprad` is encoded as `vspraw` and I'm hoping to see the `differential` fuzzer find this bug eventually with differential execution against v8.

**Comment by alexcrichton** [comment]
Well score one for fuzzing. It took a few hours but the minimal test case was 

```wasm
(module
  (type (;0;) (func (param v128 i32) (result v128)))
  (func (;0;) (type 0) (param v128 i32) (result v128)
    local.get 0
    local.get 1
    i32x4.shr_s
  )
  (export "test" (func 0))
)
```

where the differential difference was between wasmtime with avx and wasmtime without avx. That's exactly the bug I wanted the fuzzer to find, so yay confidence that fuzzing can find real bugs! I'll still let it run overnight to make sure nothing else crops up.

**Pull request merged**
